doc/source/ray-core/tasks/generators.rst

```diff
@@ -0,0 +1,78 @@
+.. _generators:
+
+Generators
+==========
+
+Python generators are functions that behave like an iterator, yielding one
+value per iteration. Ray supports remote generators for two use cases:
+
+1. To reduce max heap memory usage when returning multiple values from a remote
+   function. See the :ref:`design pattern guide <generator-pattern>` for an
+   example.
+2. When the number of return values is set dynamically by the remote function
+   instead of by the caller.
+
+
+`num_returns` set by the task caller
+------------------------------------
+
+Where possible, the caller should set the remote function's number of return values using ``@ray.remote(num_returns=x)`` or ``foo.options(num_returns=x).remote()``.
+Ray will return this many ``ObjectRefs`` to the caller.
+The remote task should then return the same number of values, usually as a tuple or list.
+
+Without changing the caller's syntax, we can also use a remote generator function to return the values iteratively.
+The generator should return the same number of return values specified by the caller, and these will be stored one at a time in Ray's object store.
+An error will be returned for generators that return a different number of values from the one specified by the caller.
+
+For example, we can swap the following code that returns a list of return values:
+
+.. literalinclude:: ../doc_code/pattern_generators.py
+    :language: python
+    :start-after: __large_values_start__
+    :end-before: __large_values_end__
+
+for this code, which uses a generator function:
+
+.. literalinclude:: ../doc_code/pattern_generators.py
+    :language: python
+    :start-after: __large_values_generator_start__
+    :end-before: __large_values_generator_end__
+
+The advantage of doing so is that the generator function does not need to hold all of its return values in memory at once.
+It can return the arrays one at a time to reduce memory pressure.
+
+`num_returns` set by the task executor
+--------------------------------------
+
+In some cases, the caller may not know the number of return values to expect from a remote function.
+For example, suppose we want to write a task that breaks up its argument into equal-size chunks and returns these.
+We may not know the size of the argument until we execute the task, so we don't know the number of return values to expect.
+
+In these cases, we can use a remote generator function that returns a *dynamic* number of values.
+To use this feature, set ``num_returns="dynamic"`` in the ``@ray.remote`` decorator or the remote function's ``.options()``.
+Then, when invoking the remote function, Ray will return an ``ObjectRefGenerator`` instead of one or multiple ``ObjectRefs``.
+
+Note that when an ``ObjectRefGenerator`` is first iterated upon, *the caller will block until the corresponding task is complete* and the number of return values is known.
```

Comment:
Hmm, I thought the return value of dynamic generator task is also an `ObjectRef` that we can get the generator by calling `ray.get()`

Reply:
This was a change that I made to try to make the semantics closer to the usual list of ObjectRefs.

The main thing that might be a bit weird is that there's an implicit ray.get when you try to access the iterator. But it's fine to pass it around to other tasks before then.

Reply:
I preferred this syntax over returning an `ObjectRef` directly, though, since it makes it explicit that a `num_returns="dynamic"` task does not return a normal `ObjectRef`.

Reply:
Isn't this limiting though? For example, if you wanted to wait for two generators to be completed, now you couldn't use ray.wait.

Reply:
You could still use `ray.wait` like this as long as we support passing ObjectRefGenerators too:
```python
g1 = gen.remote()
g2 = gen.remote()
gs, _ = ray.wait([g1, g2], num_returns=2)  # Waits until the ref lists for both are ready.
```

I think the only thing that wouldn't be supported is if you wanted to block on the results of both at the same time with the list syntax:
```python
zipped = zip(g1, g2)
for pair in zipped:
  # First iteration would wait for g1 ref list, then for g2 ref list.
  pass
```

Reply:
This has implications for other APIs as well right? (e.g., ray.kill, passing as args). It seems at the least this would change the type signature to Union[ObjectRef, ObjectRefGenerator] for many APIs.